### PR TITLE
Reuse solver residual in calibrator

### LINKED ISSUE #153: Calibrator: Reuse residual from solver
### Problem

We currently reimplement the residual in the calibrator, e.g., for the `BloodVessel` block:
https://github.com/SimVascular/svZeroDSolver/blob/32b46d7ba376c8384842902aadc45aeb6d875bbd/src/model/BloodVessel.cpp#L126-L130
However, the residual is the same as in the solver. Compare equations (2.18) and (2.24) in [this paper](https://onlinelibrary.wiley.com/doi/full/10.1002/cnm.3639). Only the Jacobian is different and needs to be implemented separately for the solver and calibrator.

### Solution

Automatically reuse the solver's residual in the calibrator for all blocks.

### Additional context

_No response_

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct and Contributing Guidelines

### PR DESCRIPTION
**Pull request by mrp089**
## Current situation
Closes #153.

## Release Notes 
The calibrator reuses the code from the solver to calculate the residual, which previously needed to be reimplemented for each block. This makes the code more straightforward and makes calibrating new blocks easier.

## Documentation
No change in functionality

## Testing
Should work as before

## Code of Conduct & Contributing Guidelines 
- [x] I agree to follow the [Code of Conduct](https://github.com/SimVascular/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/SimVascular/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mrp089** [comment]
@federicaninno, can you review this?

**Comment by federicaninno** [comment]
@mrp089 sure, will do!

File: src/optimize/LevenbergMarquardtOptimizer.cpp [hunk 1]
```diff
@@ -26,6 +26,11 @@ LevenbergMarquardtOptimizer::LevenbergMarquardtOptimizer(
   mat = Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>(num_active,
                                                               num_active);
   vec = Eigen::Matrix<double, Eigen::Dynamic, 1>::Zero(num_active);
+
+  // Set up the solver system so that its residual assembly can be reused for a
+  // single observation at a time.
+  system = SparseSystem(num_vars);
```
Comment (federicaninno) [inline src/optimize/LevenbergMarquardtOptimizer.cpp:32]:
```suggestion
 system = SparseSystem(std::max(num_vars, num_eqns));
```

The calibrator's model omits boundary condition blocks (their values come from the observations instead), so num_eqns and num_vars need not match here as they do for a full simulation. Size the system by the larger of the two so that both equation-row and variable-column indices fit.

File: src/optimize/LevenbergMarquardtOptimizer.cpp [hunk 2]
```diff
@@ -68,14 +73,46 @@ void LevenbergMarquardtOptimizer::update_gradient(
...
+      y_dpoint[k] = y_obs[i][k];
+      dy_dpoint[k] = dy_obs[i][k];
+    }
+
+    // Reuse the residual of the solver: r = -C - E*ydot - F*y. This is the
+    // same residual the solver assembles, so it does not need to be redefined
+    // here.
+    model->update_solution(system, y_dpoint, dy_dpoint);
+    system.update_residual(y_dpoint, dy_dpoint);
+    residual.segment(num_eqns * i, num_eqns) = system.residual;
```
Comment (federicaninno) [inline src/optimize/LevenbergMarquardtOptimizer.cpp]:
```suggestion
   residual.segment(num_eqns * i, num_eqns) = system.residual.head(num_eqns);
```
system.residual has max(num_vars, num_eqns) slots (see comment at line 32), which is num_vars in the calibrator, since there are no boundary condition blocks. However, only the first num_eqns slots ever get filled in, the rest stay zero. .head(num_eqns) keeps just those first num_eqns slots so the result fits into the num_eqns-sized segment.

**Review by federicaninno — CHANGES_REQUESTED** [review]
Hi @mrp089, I tried running the calibrator on this branch against the existing test case (steadyFlow_calibration.json), and I got this error:

"Assertion failed: (rows == this->rows() && cols == this->cols() && "DenseBase::resize() does not actually allow to resize."), function resize, file DenseBase.h, line 262.
zsh: abort      ./svzerodcalibrator ../tests/cases/steadyFlow_calibration.json try.json"

It seems like "calibrate.cpp" only builds vessel/junction blocks and skips the boundary-condition blocks (FlowReferenceBC, ResistanceBC, etc.), since those values come straight from the observation data instead of being solved for. But those BC blocks are normally what add the extra equations that make the system square (num_vars == num_eqns). Without them, we end up with more variables than equations, which then creates problems in "LevenbergMarquardtOptimizer.cpp". At line 32: system = SparseSystem(num_vars); sizes system.residual to num_vars. Later, at line 102: residual.segment(num_eqns * i, num_eqns) = system.residual tries to stuff that num_vars-sized vector into a num_eqns-sized slot, and cannot resize to fit, hence the crash. I am leaving a possible fix directly as comments on the two lines in question, which I tested and worked.

Let me know what you think!

**Comment by mrp089** [comment]
Well caught, @federicaninno! Turns out our CI pipeline still passed when `Eigen` threw an assertion error. This got fixed in [7fa4755](https://github.com/SimVascular/svZeroDSolver/pull/231/commits/7fa4755aab435d3a7793fc92e5bd63ad2e2c7250) with the pipelines failing accordingly. The actual code fix is in [e23fdc0](https://github.com/SimVascular/svZeroDSolver/pull/231/commits/e23fdc0240719e2ed6c8460aa06232c574573dd7). Ready for your re-review!

**Review by federicaninno — APPROVED** [review]

**Comment by federicaninno** [comment]
All good for me! Merging this branch to master
